pkg/notes/notes_v2.go

```diff
@@ -0,0 +1,249 @@
+		}
+	}
+
+	if err := t.Err(); err != nil {
+		return nil, err
+	}
+
+	bar.Finish()
+
+	return aggregator.releaseNotes, nil
+}
+
+func (g *Gatherer) buildReleaseNote(pair *commitPrPair) (*ReleaseNote, error) {
+	pr, _, err := g.client.GetPullRequest(g.context, g.options.GithubOrg, g.options.GithubRepo, pair.PrNum)
+	if err != nil {
+		return nil, err
+	}
+
+	prBody := pr.GetBody()
+
+	text, err := noteTextFromString(prBody)
+	if err != nil {
+		logrus.Debugf("sha: %s pr: %d err: %v", pair.Commit.Hash.String(), pair.PrNum, err)
+		return nil, nil
+	}
+
+	documentation := DocumentationFromString(prBody)
+
+	author := pr.GetUser().GetLogin()
+	authorURL := pr.GetUser().GetHTMLURL()
+	prURL := pr.GetHTMLURL()
+	isFeature := hasString(labelsWithPrefix(pr, "kind"), "feature")
+	noteSuffix := prettifySIGList(labelsWithPrefix(pr, "sig"))
+
+	isDuplicateSIG := false
+	if len(labelsWithPrefix(pr, "sig")) > 1 {
+		isDuplicateSIG = true
+	}
+
+	isDuplicateKind := false
+	if len(labelsWithPrefix(pr, "kind")) > 1 {
+		isDuplicateKind = true
+	}
+
+	// TODO(wilsonehusin): extract / follow original in ReleasenoteFromCommit
+	indented := strings.ReplaceAll(text, "\n", "\n  ")
+	markdown := fmt.Sprintf("%s ([#%d](%s), [@%s](%s))",
+		indented, pr.GetNumber(), prURL, author, authorURL)
+
+	if noteSuffix != "" {
+		markdown = fmt.Sprintf("%s [%s]", markdown, noteSuffix)
+	}
+
+	// Uppercase the first character of the markdown to make it look uniform
+	markdown = strings.ToUpper(string(markdown[0])) + markdown[1:]
+
+	return &ReleaseNote{
+		Commit:         pair.Commit.Hash.String(),
+		Text:           text,
+		Markdown:       markdown,
+		Documentation:  documentation,
+		Author:         author,
+		AuthorURL:      authorURL,
+		PrURL:          prURL,
+		PrNumber:       pr.GetNumber(),
+		SIGs:           labelsWithPrefix(pr, "sig"),
+		Kinds:          labelsWithPrefix(pr, "kind"),
+		Areas:          labelsWithPrefix(pr, "area"),
+		Feature:        isFeature,
+		Duplicate:      isDuplicateSIG,
+		DuplicateKind:  isDuplicateKind,
+		ActionRequired: labelExactMatch(pr, "release-note-action-required"),
+		DoNotPublish:   labelExactMatch(pr, "release-note-none"),
+	}, nil
+}
+
+func (g *Gatherer) listLeftParentCommits(opts *options.Options) ([]*commitPrPair, error) {
+	localRepository, err := git.PlainOpen(opts.RepoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// opts.StartSHA points to a tag (e.g. 1.20.0) which is on a release branch (e.g. release-1.20)
+	// this means traveling through commit history from opts.EndSHA will never reach opts.StartSHA
+
+	// the stopping point to be set should be the last shared commit between release branch and primary (master) branch
+	// usually, following the left / first parents, it would be
+	// * tag: v1.20.0, some merge commit
+	// |
+	// * Anago GCB release commit (begin branch out of release-1.20)
+	// |
+	// * last shared commit
+
+	// this means the stopping point is 2 commits behind the tag pointed by opts.StartSHA
```

Comment:
Can we ensure that we there are always only 2 commits? This would change if we decide to cherry-pick some patches right before the release, right?

Reply:
That's great, thank you! :pray:

Reply:
so I looked into this again and I think trying to support patch releases at the moment would over-complicate the code. for example, in generating patch releases of 1.20, we need to find commits from the specified tag up to 1.19.0, which points to [this commit](https://github.com/kubernetes/kubernetes/commit/faecb196815e248d3ecfb03c680a4507229c2a56). somehow, 2-level of parents from that commit has release-1.19 on the left and master on the right, which is the opposite of the convention in 1.20.

in short, I'm suggesting is that we leave this `--list-v2` experimental flag until 1.20 is the oldest release line we support before deprecating the older implementation. `--list-v2` will continue to work and beneficial for release team fetching minor-release notes and patch-release notes won't even get a chance to use this until later. how does that sound?

cc @puerco @saschagrunert

Reply:
Well, we could always plug the v2 list code into the `krel release-notes ...`  invocations used by the release team. 

Also, let's keep in mind that it's not just a question of kubernetes prereleases vs patch releases. We already have 3rd party consumers who are using the release-notes tool for other projects. How are those looking facing a potential change? Can we know?

Reply:
so I don't expect `--list-v2` implementation to change the end result. it _fixed a bug_ which exists the old implementation, and that's the only acceptable regression IMO. anything else has to be the same, just performance improvements.

> We already have 3rd party consumers who are using the release-notes tool for other projects. How are those looking facing a potential change? Can we know?

I think this is the tricky part. they are safe for now because of the feature flag, but I wonder if they rely on the bug mentioned above for their work. do we know more concrete consumers for this?

Reply:
okay I went ahead and implemented commit resolution through `merge-base` and it's now working for patch releases! :tada: